src/machine_charm.py

```diff
@@ -140,29 +140,44 @@ def _get_principal_relation(self) -> Optional[Relation]:
             return self.model.relations["juju-info"][0]
         else:
             return None
-
+        
+    @property
+    def principal_unit(self):
```

Comment:
instead of a property, I'd make this an instance attribute. Set it in `__init__` and be done with it, no? it's not going to change later. Also `is_ready`, by the way.